### ISSUE TITLE
Filter entries in the logger by filterlist

### ISSUE THREAD

**Issue by peace2000** [uBlockOrigin/uBlock-issues#842]
Hi! Can you add an option to filter entries in the logger, based on filterlist?

This would very be helpful if one needs to check if a specific filterlist or filterlists have blocked or whitelisted entries on a specific website. I know that I can click each blocked / whitelisted entry in the logger and see which list blocked/whitelisted it but clicking through many entries might be slow and laborious task in some cases. Especially if many filterlists are being used and if a site produces many entries in the logger.

If I could just filter out some lists in the logger it would speed up things and reduce unnecessary manual work. Also I would not need to disable / enable different lists in the Ubo settings in order to rule out which list(s) produces entries.

**Comment by gorhill** [comment]
Finding the filter list(s) in which a filter is found is an expensive operation CPU-wise, hence why it's done on-demand only when clicking on a row for more details.

**Comment by peace2000** [comment]
But if it's already possible to click a blocked entry open and see which filter blocked it, how come it would cpu intensive to filter entries according to that data that is already available?

https://user-images.githubusercontent.com/886325/62740615-97d89800-ba38-11e9-9b4c-869fac0d0129.png

![kuva](https://user-images.githubusercontent.com/17256841/71847796-90e0af00-30d6-11ea-8be6-cf2db5d3770d.png)

**Comment by gorhill** [comment]
> how come it would cpu intensive to filter entries according to that data that is already available?

I said:

> it's done on-demand only when clicking on a row for more details.

You are asking that uBO does the lookup at once for all entries in the logger for which there is a filter. There can be tens, hundreds or even more such entries -- so essentially multiplying by 10, 100 or more the cost of looking up filter list(s) from one filter.

**Comment by peace2000** [comment]
Got it now.

Well, I don't know how much cpu power it would in practice take, but some applications that have cpu intensive features or functions, issue a warning text "this feature could use lots of cpu" etc.

Maybe that function could be added for those who have a powerful cpu?

But anyway, it's always developer's decision :) But this kind of functionality would help a lot.

**Comment by mapx-** [comment]
Maybe the information (belonging to a list) is lost when compiling / validating filters.

Is it possible to concatenate this requested information (for example by encoding ublock filters = 1, easylist = 2 etc.) while compiling the filters ?

**Comment by uBlock-user** [comment]
> how come it would cpu intensive to filter entries according to that data that is already available?

The data is not already available, once you click on the filter, the reverse lookup happens, but the operation is so fast performance-wise, that you won't even notice any delay/lag when looking for the filterlist name.

You can serve this info without the performance hit, if there was a way to retreive the info without having to use reverse-lookup.

**Comment by peace2000** [comment]
> > how come it would cpu intensive to filter entries according to that data that is already available?
> 
> The data is not already available, once you click on the filter, the reverse lookup happens, but the operation is so fast performance-wise, that you won't even notice any delay/lag when looking for the filterlist name.
> 
> You can serve this info without the performance hit, if there was a way to retreive the info without having to use reverse-lookup.

Good clarification.

**Comment by peace2000** [comment]
> 
> 
> Maybe the information (filter belonging to a specific list) is lost during compiling / validating filters process.
> 
> Is it possible to concatenate (to the filter) this requested information (for example by encoding ublock filters = 1, easylist = 2 etc.) while compiling the filters ?

I happened to notice that Adguard does have this functionality:

![kuva](https://user-images.githubusercontent.com/17256841/71995057-94914480-3242-11ea-8ca6-e3310178daa4.png)

Here is one way how this could be implemented in the logger.

**Comment by kulfoon** [comment]
Related: https://github.com/uBlockOrigin/uBlock-issues/issues/207

**Comment by Yuki2718** [comment]
> 
> I happened to notice that Adguard does have this functionality:
> 

However, you can't filter the results out by lists, and it only shows one filter even when the entry hits 2+ lists.

**Comment by peace2000** [comment]
@Yuki2718 Yes to exact, the point was just that Adguard shows in the logger which lists have hits. I noticed also that filtering by list is not implemented.

"it only shows one list even when the entry hits 2+ lists."

Well this is clearly either a bug or a design flaw.

**Comment by gorhill** [comment]
I don't need to be shown a static screenshot of AdGuard's logger, surely it should be safely assumed I have a clue about how to implement a filter list(s) column if I ever choose to do so after years working on these extensions.

The reason why it's not like this in uBO is strictly because efficiency is a key feature in uBO and choices were made architecturally to ensure filters are efficiently stored and those choices meant to give up `[ filter => filter list(s) ]` associations, hence why it's resolved on demand through user interaction from within the logger. If you keep you eyes on uBO's memory usage you will notice a large jump in memory usage when you lookup filter list(s) from a filter the first time.

Now, I won't sacrifice uBO's efficiency, so the only sensible approach is to batch process the reverse lookup of filter list(s) to build the dictionary on the fly in the logger code. Leaving aside worries about efficiency, that alone is not trivial, and there are other issues regarding keeping adding code to the logger and frankly I don't have the motivation to keep complicating the logger's code for a feature which I consider a edge case, not a must-have feature.

**Comment by peace2000** [comment]
> 
> 
> I don't need to be shown a static screenshot of AdGuard's logger, surely it should be safely assumed I have a clue about how to implement a filter list(s) column if I ever choose to do so after years working on these extensions.

I don't doubt that. :) It was just kind of reference picture to support this issue and what kind of things I had in my mind.

> The reason why it's not like this in uBO is strictly because efficiency is a key feature in uBO and choices were made architecturally to ensure filters are efficiently stored and those choices meant to give up `[ filter => filter list(s) ]` associations, hence why it's resolved on demand through user interaction from within the logger. If you keep you eyes on uBO's memory usage you will notice a large jump in memory usage when you lookup filter list(s) from a filter the first time.
> 
> Now, I won't sacrifice uBO's efficiency, so the only sensible approach is to batch process the reverse lookup of filter list(s) to build the dictionary on the fly in the logger code. Leaving aside worries about efficiency, that alone is not trivial, and there are other issues regarding keeping adding code to the logger and frankly I don't have the motivation to keep complicating the logger's code for a feature which I consider a edge case, not a must-have feature.

I appreciate your efficiency based approach. That's why I later suggested an on-off switch based approach.
